openml/datasets/functions.py

```diff
@@ -353,6 +353,39 @@ def get_dataset(dataset_id):
     return dataset
 
 
+def _pandas_dtype_to_arff_dtype(df, column_name, column_dtype):
+    """Map Pandas dtype to ARFF dtype
+
+    Arguments:
+        df : DataFrame, shape (n_samples, n_features)
+            The dataframe containing the data.
+        column_name : str
+            The name of the column for which we want to infer the dtype.
+        column_dtype : dtype
+            The Pandas data type of the column of interest.
+
+    Returns:
+        arff_dtype : str
+            The ARFF dtype for the given column.
+    """
+    if column_dtype.name == 'category':
+        arff_dtype = df[column_name].unique().tolist()
+        # for categorical feature, arff expects a list string. However, a
+        # categorical column can contain mixed type and we should therefore
+        # raise an error asking to convert all enties to string.
+        if not all([isinstance(cat, six.string_types) for cat in arff_dtype]):
+            raise ValueError("The column '{}' of the dataframe is of "
+                             "'category' dtype. Therefore, all values in this "
+                             "columns should be string. Please convert the "
+                             "entries which are not string."
+                             .format(column_name))
+    elif column_dtype.name == 'object':
+        arff_dtype = 'STRING'
```

Comment:
Usually we want this to be category as well, right? kinda tricky ;)
And what about integers that encode categories?

Reply:
> And what about integers that encode categories?

If they encode categories, then one should use the `category` dtype for those columns.

However, this is part needs to be modified. We should manage `INTEGER` and `REAL`/`NUMERIC` differently. We also have to think more in depth about the missing values. I'll use the expertise of @jorisvandenbossche

Basically, we are working on uploading the datasets from `dirty_cat` and we are confronted to those issues.

Reply:
Yes, I would also say that it is up to the user to use a categorical dtype. Actual string columns are not that rare (eg titanic dataset), and it is quite impossible to guess from a string column if it should be categorical or not. 

Of course, it could also be a keyword that the user can pass to indicate which columns to see as categorical without the need that the (s)he does the actual conversion themselves.

>  And what about integers that encode categories?

I think arff does not support integer categories?

Reply:
Conceptually, arff should allow this (as long as the categories are integers, too). However, liac-arff does not support this. Can be reproduced with:
```
import arff

arff.dumps({
    'relation': 'test',
    'description': 'abc',
    'attributes': [['test', [1, 2, 3]]],
    'data': [[1], [2]],
})
```
I'm wondering if it's worth changing this. Formally, there's also no difference between REAL, INTEGER and NUMERIC, but I don't see a reason to not be more specific about the data type as the arff specification allows this.

Reply:
I think that there is a difference between REAL and INTEGER since the dtype will be kept, isn't it.

```python
In [5]: X = arff.dumps({ 
   ...:   'relation': 'test', 
   ...:   'description': 'abc', 
   ...:   'attributes': [['test', 'REAL']], 
   ...:   'data': [[1], [2]] 
   ...: })                                                                                              

In [6]: arff.loads(X)                                                                                   
Out[6]: 
{'description': 'abc',
 'relation': 'test',
 'attributes': [('test', 'REAL')],
 'data': [[1.0], [2.0]]}

In [7]: X = arff.dumps({ 
   ...:   'relation': 'test', 
   ...:   'description': 'abc', 
   ...:   'attributes': [['test', 'INTEGER']], 
   ...:   'data': [[1], [2]] 
   ...: })                                                                                              

In [8]: arff.loads(X)                                                                                   
Out[8]: 
{'description': 'abc',
 'relation': 'test',
 'attributes': [('test', 'INTEGER')],
 'data': [[1], [2]]}
```

Reply:
Going by the official arff specification, all numeric attributes should go by NUMERIC 
https://www.cs.waikato.ac.nz/ml/weka/arff.html

In order to make OpenML datasets comply with as much as possible arff parsers, maybe it's best to enforce this too server side? In that case the Python uploader should of course also upload NUMERIC key. @mfeurer @joaquinvanschoren @berndbischl WDYT?

Reply:
Have to add that currently this is not enforced, server side

Reply:
@janvanrijn you pointed to an outdated version of the arff standard (see the first sentence on the page you linked to), the newer one (https://waikato.github.io/weka-wiki/arff_stable/) actually mentions 'integer' and 'real' (although it says they are treated as numeric)